src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/HalfConverter.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    internal sealed class HalfConverter : JsonPrimitiveConverter<Half>
+    {
+        private const int MaxFormatLength = 16;
+        private const int MaxEscapedFormatLength = MaxFormatLength * JsonConstants.MaxExpansionFactorWhileEscaping;
+
+        public HalfConverter()
+        {
+            IsInternalConverterForNumberType = true;
+        }
+
+        public override Half Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            if (reader.TokenType != JsonTokenType.Number)
+            {
+                ThrowHelper.ThrowInvalidOperationException_ExpectedNumber(reader.TokenType);
+            }
+
+            return ReadCore(ref reader);
+        }
+
+        public override void Write(Utf8JsonWriter writer, Half value, JsonSerializerOptions options)
+        {
+            WriteCore(writer, value);
+        }
+
+        private static Half ReadCore(ref Utf8JsonReader reader)
+        {
+            Half result;
+
+            if (reader.ValueLength > MaxEscapedFormatLength)
+            {
+                ThrowHelper.ThrowFormatException(NumericType.Half);
+            }
+
+            Span<byte> buffer = stackalloc byte[MaxFormatLength];
+            int written = reader.CopyString(buffer);
+            buffer = buffer.Slice(0, written);
+
+            if (reader.TokenType is JsonTokenType.String or JsonTokenType.PropertyName)
+            {
+                if (JsonReaderHelper.TryGetFloatingPointConstant(buffer, out result))
+                {
+                    return result;
+                }
+            }
+
+#if NET8_0_OR_GREATER
+            bool success = Half.TryParse(buffer, out result);
+#else
+            Span<char> charBuffer = stackalloc char[MaxFormatLength];
```

Comment:
I wonder if much of the complexity in this converter might be eliminated if we simply converted the `Half` value to `float` or `double` when serializing and vice versa when deserializing, using the existing implementation for these types. 

@tannergooding might be able to clarify if there are any gotchas in such an approach.

Reply:
That could potentially lead to incorrect results due to double rounding caused by the different domain boundaries. Someone would need to sit down and do the math to determine if it was definitively safe or not, and that's likely significantly more effort.